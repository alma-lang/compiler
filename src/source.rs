use crate::ast::ModuleName;
use std::cmp::{max, min};
use std::fmt;
use std::fs;
use std::io;
use std::ops::Range;
use std::path::{Path, PathBuf};
use std::slice::SliceIndex;
use std::str::CharIndices;

pub enum Error {
    InvalidExtension(String),
    InvalidFileName(String),
    IO(io::Error),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use Error::*;
        match self {
            InvalidExtension(extension) => write!(
                f,
                "Invalid file extension '{extension}'. Please use the '.alma' extension",
            ),
            InvalidFileName(name) => write!(
                f,
                "Invalid file name '{name}'. Alma file names must look like this 'MyFile.alma'",
            ),
            IO(err) => write!(f, "There was an error reading the file.\n\n{err}"),
        }
    }
}

#[derive(Debug, PartialEq)]
enum SourceOrigin {
    File(PathBuf),
    NotAFile,
}

type LinesAround<'str> = (Vec<&'str str>, (usize, Vec<&'str str>), Vec<&'str str>);

#[derive(Debug, PartialEq)]
pub struct Source {
    source: SourceOrigin,
    code: String,
}

impl Source {
    pub fn new_file(file: String) -> Result<Self, Error> {
        let path = Path::new(&file).to_path_buf();

        Source::validate_file_name(&path)?;
        let code = Source::read_file(&path)?;

        Ok(Source {
            source: SourceOrigin::File(path),
            code,
        })
    }

    fn validate_file_name(file_path: &Path) -> Result<(), Error> {
        let file_name = file_path
            .file_stem()
            // Convert OSStr to str
            .and_then(|s| s.to_str())
            .unwrap_or_default();
        let extension = file_path
            .extension()
            // Convert OSStr to str
            .and_then(|s| s.to_str())
            .unwrap_or_default();

        if extension != "alma" {
            Err(Error::InvalidExtension(extension.to_string()))
        } else if !ModuleName::valid_part(file_name) {
            Err(Error::InvalidFileName(file_name.to_string()))
        } else {
            Ok(())
        }
    }

    fn read_file(file_path: &Path) -> Result<String, Error> {
        fs::read_to_string(file_path).map_err(Error::IO)
    }

    pub fn new_orphan(code: String) -> Self {
        Source {
            source: SourceOrigin::NotAFile,
            code,
        }
    }

    pub fn name(&self) -> &str {
        match &self.source {
            SourceOrigin::File(path) => path.to_str().unwrap(),
            SourceOrigin::NotAFile => "",
        }
    }

    pub fn file_stem(&self) -> Option<&str> {
        match &self.source {
            SourceOrigin::File(path) => Some(path.file_stem().unwrap().to_str().unwrap()),
            SourceOrigin::NotAFile => None,
        }
    }

    pub fn _char_at(&self, i: usize) -> Option<char> {
        self.code.get(i..).and_then(|code| code.chars().next())
    }

    pub fn text_at<I>(&self, i: I) -> Option<&<I as SliceIndex<str>>::Output>
    where
        I: SliceIndex<str>,
    {
        self.code.get(i)
    }

    fn _line_at(&self, position: usize) -> Option<&str> {
        self.code.get(line_at(&self.code, position)?)
    }

    pub fn lines_around_position(
        &self,
        position: usize,
        end_position: Option<usize>,
        number_of_lines: u32,
    ) -> Option<LinesAround> {
        lines_around_position(&self.code, position, end_position, number_of_lines)
    }

    pub fn len(&self) -> usize {
        self.code.len()
    }

    pub fn char_indices(&self) -> CharIndices {
        self.code.char_indices()
    }

    pub fn lines_report_at_position(
        &self,
        position: usize,
        end_position: Option<usize>,
        line_number: u32,
        show_pointer: bool,
    ) -> Option<String> {
        if self.len() == 0 {
            return None;
        }

        let mut message = String::new();

        let point_to_end_of_input = position == self.len();
        let (position, end_position) = if point_to_end_of_input {
            (position - 1, None)
        } else {
            (position, end_position)
        };

        let (lines_before, (lines_start_position, lines), lines_after) =
            self.lines_around_position(position, end_position, 2)?;

        let line_number_width: usize = (line_number as usize + lines.len() + lines_after.len())
            .to_string()
            .len();

        let column_number = position - lines_start_position;

        for (i, l) in lines_before.iter().enumerate() {
            if !message.is_empty() {
                message.push('\n');
            }
            message.push_str(&format!(
                "  {line_number:line_number_width$}│  {l}",
                line_number = line_number - lines_before.len() as u32 + i as u32,
            ));
        }

        match lines.as_slice() {
            [] => panic!("Empty code for report, should've bailed earlier at line_at"),
            [line] => {
                if !message.is_empty() {
                    message.push('\n');
                }
                message.push_str(&format!("  {line_number:line_number_width$}│  {line}\n"));

                if show_pointer {
                    let num_pointers = max(1, end_position.unwrap_or(position) - position);

                    let num_spaces = if point_to_end_of_input {
                        column_number + 1
                    } else {
                        column_number
                    };
                    // In the case input ends with \n and we point to it, num_spaces here will end
                    // up being off the line by 1 position. Instead of checking somehow if it ended
                    // in \n, and it was point_to_end_of_input, just cap the pointer to point to
                    // the end of line at the furthest.
                    let num_spaces = min(line.len(), num_spaces);

                    message.push_str(&format!(
                        "  {space:line_number_width$}│  {spaces}{pointers}",
                        space = ' ',
                        spaces = str::repeat(" ", num_spaces as usize),
                        pointers = str::repeat("↑", num_pointers as usize)
                    ));
                }
            }
            lines => {
                for (i, l) in lines.iter().enumerate() {
                    if !message.is_empty() {
                        message.push('\n');
                    }
                    message.push_str(&format!(
                        "  {line_number:line_number_width$}│{line_marker} {l}",
                        line_number = line_number + i as u32,
                        line_marker = if show_pointer { "→" } else { " " },
                    ));
                }
            }
        }

        for (i, l) in lines_after.iter().enumerate() {
            message.push('\n');
            message.push_str(&format!(
                "  {line_number:line_number_width$}│  {l}",
                line_number = line_number + 1 + i as u32,
            ));
        }
        Some(message)
    }

    pub fn lines_report_at_position_with_pointer(
        &self,
        position: usize,
        end_position: Option<usize>,
        line_number: u32,
    ) -> Option<String> {
        self.lines_report_at_position(position, end_position, line_number, true)
    }

    pub fn to_string_with_line_and_col(&self, line: u32, column: u32) -> String {
        let file_name = match &self.source {
            SourceOrigin::File(path) => path.to_str().unwrap(),
            SourceOrigin::NotAFile => "",
        };
        let space = if file_name.is_empty() { "" } else { " " };
        format!("{file_name}{space}[{line}:{column}]")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    fn report(code: &str, position: usize, end_position: Option<usize>, lines: u32) -> String {
        let source = Source::new_orphan(code.to_string());
        let result = source.lines_report_at_position_with_pointer(position, end_position, lines);
        format!(
            "Input:\n\n{code}\n\nResult:\n\n{result}",
            result = match &result {
                Some(res) => res,
                None => "None",
            }
        )
    }

    #[test]
    fn test_pointing_to_the_end_of_file_points_to_last_character_even_with_new_lines() {
        assert_snapshot!(report("type Banana\n", 12, None, 1));
    }
}

fn line_at(code: &str, position: usize) -> Option<Range<usize>> {
    if position >= code.len() {
        return None;
    }

    let line_start = {
        let back = code.get(..position)?;
        if back.is_empty() {
            position
        } else {
            back.char_indices()
                .rfind(|(i, c)| *c == '\n' || *i == 0)
                .map(|(i, c)| if i == 0 && c != '\n' { i } else { i + 1 })?
        }
    };

    // Exclusive end index
    let line_end = {
        let front = code.get(position..)?;
        front
            .char_indices()
            .find(|(i, c)| *c == '\n' || *i == front.len() - 1)
            .map(|(i, c)| {
                (if i == front.len() - 1 && c != '\n' {
                    i + 1
                } else {
                    i
                }) + position
            })?
    };

    Some(line_start..line_end)
}

#[test]
fn test_line_at() {
    let tests = vec![
        ("", 0, None),
        ("\n", 0, Some("")),
        ("\n", 1, None),
        ("\n\n", 0, Some("")),
        ("\n\n", 1, Some("")),
        ("a", 0, Some("a")),
        ("a\n", 1, Some("a")),
        ("\na", 1, Some("a")),
        ("\n\na", 1, Some("")),
        ("\n\na", 2, Some("a")),
        ("\n\na\n", 2, Some("a")),
        ("hello\nhi\nbye", 0, Some("hello")),
        ("hello\nhi\nbye", 2, Some("hello")),
        ("hello\nhi\nbye", 4, Some("hello")),
        ("hello\nhi\nbye", 5, Some("hello")),
        ("hello\nhi\nbye", 6, Some("hi")),
        ("hello\nhi\nbye", 7, Some("hi")),
        ("hello\nhi\nbye", 8, Some("hi")),
        ("hello\nhi\nbye", 9, Some("bye")),
        ("hello\nhi\nbye", 10, Some("bye")),
        ("hello\nhi\nbye", 11, Some("bye")),
    ];

    for (s, i, expected) in tests {
        let line = line_at(s, i).and_then(|i| s.get(i));
        assert_eq!(line, expected, "text = {:?}, get line at char {}", s, i);
    }
}

fn lines_before_position(code: &str, position: usize, number_of_lines: u32) -> Option<Vec<&str>> {
    let mut lines = Vec::new();
    let mut how_many_lines = number_of_lines;

    if position > 0 && how_many_lines > 0 {
        let mut idx = position - 1;
        while let Some(bounds) = line_at(code, idx) {
            let start = bounds.start;
            lines.push(code.get(bounds)?);
            how_many_lines -= 1;
            if start == 0 || how_many_lines == 0 {
                break;
            }
            idx = start - 1;
        }
        lines.reverse();
    }

    Some(lines)
}

fn lines_after_position(code: &str, position: usize, number_of_lines: u32) -> Option<Vec<&str>> {
    let mut lines = Vec::new();
    let mut how_many_lines = number_of_lines;

    if position < code.len() - 1 && how_many_lines > 0 {
        let mut idx = position + 1;
        while let Some(bounds) = line_at(code, idx) {
            let end = bounds.end;
            lines.push(code.get(bounds)?);
            how_many_lines -= 1;
            if end >= code.len() - 1 || how_many_lines == 0 {
                break;
            }
            idx = end + 1;
        }
    }

    Some(lines)
}

fn lines_around_position(
    code: &str,
    position: usize,
    end_position: Option<usize>,
    number_of_lines: u32,
) -> Option<LinesAround> {
    let line_bounds = line_at(code, position)?;

    match end_position {
        Some(end) => {
            // end is exclusive, so substract 1 to get the line at the actual last character
            let end_line_bounds = line_at(code, end - 1)?;

            let lines = code
                .get(line_bounds.start..end_line_bounds.end)?
                .lines()
                .collect();

            let lines_before = lines_before_position(code, line_bounds.start, number_of_lines)?;

            let lines_after = lines_after_position(code, end_line_bounds.end, number_of_lines)?;

            Some((lines_before, (line_bounds.start, lines), lines_after))
        }
        None => {
            let line = code.get(line_bounds.clone())?;

            let lines_before = lines_before_position(code, line_bounds.start, number_of_lines)?;

            let lines_after = lines_after_position(code, line_bounds.end, number_of_lines)?;

            Some((lines_before, (line_bounds.start, vec![line]), lines_after))
        }
    }
}

#[test]
fn test_lines_around_position() {
    let tests = vec![
        ("", 0, None, 1, None),
        ("\n", 0, None, 1, Some((vec![], (0, vec![""]), vec![]))),
        ("\n\n", 0, None, 1, Some((vec![], (0, vec![""]), vec![""]))),
        ("\n\n", 1, None, 1, Some((vec![""], (1, vec![""]), vec![]))),
        ("a", 0, None, 1, Some((vec![], (0, vec!["a"]), vec![]))),
        ("a\n", 0, None, 1, Some((vec![], (0, vec!["a"]), vec![]))),
        ("a\n", 1, None, 1, Some((vec![], (0, vec!["a"]), vec![]))),
        ("\na", 0, None, 1, Some((vec![], (0, vec![""]), vec!["a"]))),
        ("\na", 1, None, 1, Some((vec![""], (1, vec!["a"]), vec![]))),
        ("\n\na", 0, None, 1, Some((vec![], (0, vec![""]), vec![""]))),
        (
            "\n\na",
            0,
            None,
            2,
            Some((vec![], (0, vec![""]), vec!["", "a"])),
        ),
        (
            "\n\na",
            0,
            None,
            3,
            Some((vec![], (0, vec![""]), vec!["", "a"])),
        ),
        (
            "\n\na",
            1,
            None,
            1,
            Some((vec![""], (1, vec![""]), vec!["a"])),
        ),
        (
            "\n\na",
            1,
            None,
            2,
            Some((vec![""], (1, vec![""]), vec!["a"])),
        ),
        (
            "\n\na",
            1,
            None,
            3,
            Some((vec![""], (1, vec![""]), vec!["a"])),
        ),
        (
            "\n\na",
            2,
            None,
            1,
            Some((vec![""], (2, vec!["a"]), vec![])),
        ),
        (
            "\n\na",
            2,
            None,
            2,
            Some((vec!["", ""], (2, vec!["a"]), vec![])),
        ),
        (
            "\n\na",
            2,
            None,
            3,
            Some((vec!["", ""], (2, vec!["a"]), vec![])),
        ),
        (
            "hello\nhi\nbye",
            0,
            None,
            1,
            Some((vec![], (0, vec!["hello"]), vec!["hi"])),
        ),
        (
            "hello\nhi\nbye",
            0,
            None,
            2,
            Some((vec![], (0, vec!["hello"]), vec!["hi", "bye"])),
        ),
        (
            "hello\nhi\nbye",
            0,
            None,
            3,
            Some((vec![], (0, vec!["hello"]), vec!["hi", "bye"])),
        ),
        (
            "hello\nhi\nbye",
            2,
            None,
            1,
            Some((vec![], (0, vec!["hello"]), vec!["hi"])),
        ),
        (
            "hello\nhi\nbye",
            2,
            None,
            2,
            Some((vec![], (0, vec!["hello"]), vec!["hi", "bye"])),
        ),
        (
            "hello\nhi\nbye",
            2,
            None,
            3,
            Some((vec![], (0, vec!["hello"]), vec!["hi", "bye"])),
        ),
        (
            "hello\nhi\nbye",
            5,
            None,
            1,
            Some((vec![], (0, vec!["hello"]), vec!["hi"])),
        ),
        (
            "hello\nhi\nbye",
            5,
            None,
            2,
            Some((vec![], (0, vec!["hello"]), vec!["hi", "bye"])),
        ),
        (
            "hello\nhi\nbye",
            5,
            None,
            3,
            Some((vec![], (0, vec!["hello"]), vec!["hi", "bye"])),
        ),
        (
            "hello\nhi\nbye",
            6,
            None,
            1,
            Some((vec!["hello"], (6, vec!["hi"]), vec!["bye"])),
        ),
        (
            "hello\nhi\nbye",
            6,
            None,
            2,
            Some((vec!["hello"], (6, vec!["hi"]), vec!["bye"])),
        ),
        (
            "hello\nhi\nbye",
            8,
            None,
            1,
            Some((vec!["hello"], (6, vec!["hi"]), vec!["bye"])),
        ),
        (
            "hello\nhi\nbye",
            8,
            None,
            2,
            Some((vec!["hello"], (6, vec!["hi"]), vec!["bye"])),
        ),
        (
            "hello\nhi\nbye",
            9,
            None,
            1,
            Some((vec!["hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye",
            9,
            None,
            2,
            Some((vec!["hello", "hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye",
            9,
            None,
            3,
            Some((vec!["hello", "hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye",
            11,
            None,
            1,
            Some((vec!["hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye",
            11,
            None,
            2,
            Some((vec!["hello", "hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye",
            11,
            None,
            3,
            Some((vec!["hello", "hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye\n",
            12,
            None,
            1,
            Some((vec!["hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye\n",
            12,
            None,
            2,
            Some((vec!["hello", "hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "hello\nhi\nbye\n",
            12,
            None,
            3,
            Some((vec!["hello", "hi"], (9, vec!["bye"]), vec![])),
        ),
        (
            "line1\nline2\nline3\nline4\nline5\nline6",
            3,
            Some(5),
            3,
            Some((vec![], (0, vec!["line1"]), vec!["line2", "line3", "line4"])),
        ),
        (
            "line1\nline2\nline3\nline4\nline5\nline6",
            3,
            // Exclusive end. This is still line 1
            Some(6),
            3,
            Some((vec![], (0, vec!["line1"]), vec!["line2", "line3", "line4"])),
        ),
        (
            "line1\nline2\nline3\nline4\nline5\nline6",
            3,
            Some(7),
            3,
            Some((
                vec![],
                (0, vec!["line1", "line2"]),
                vec!["line3", "line4", "line5"],
            )),
        ),
        (
            "line1\nline2\nline3\nline4\nline5\nline6",
            7,
            Some(13),
            3,
            Some((
                vec!["line1"],
                (6, vec!["line2", "line3"]),
                vec!["line4", "line5", "line6"],
            )),
        ),
    ];

    for (s, position, end_position, n_lines, expected) in tests {
        let lines = lines_around_position(s, position, end_position, n_lines);
        assert_eq!(
            lines, expected,
            "text = {:?}, get {} lines around char {} and {:?}",
            s, n_lines, position, end_position
        );
    }
}
