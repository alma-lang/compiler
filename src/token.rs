use crate::strings::{self, Strings};
use derive_more::{Add, AddAssign, From, Into, Sub, SubAssign};
use typed_index_collections::TiVec;

#[derive(
    Debug, From, Into, Add, AddAssign, Sub, SubAssign, PartialEq, Eq, PartialOrd, Ord, Clone, Copy,
)]
pub struct Index(usize);

pub type Tokens = TiVec<Index, Token>;

#[derive(PartialEq, Debug, Copy, Clone)]
pub struct Token {
    pub kind: Type,
    pub indent: u32,
    pub start: usize,
    pub end: usize,
    pub line: u32,
    pub column: u32,
}

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum Type {
    // SingleToken-character tokens.
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Colon,
    Slash,
    Backslash,
    Star,
    Underscore,
    Pipe,

    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    Arrow,

    // Literals.
    Identifier(strings::Symbol),
    CapitalizedIdentifier(strings::Symbol),
    String_(strings::Symbol),
    Float(strings::Symbol),

    // Keywords.
    And,
    Or,
    Not,
    If,
    Then,
    Else,
    True,
    False,
    Let,
    In,
    As,
    Exposing,
    Module,
    Import,
    Type,

    Comment,
    Eof,
}
impl Type {
    pub fn to_string<'strings>(&self, strings: &'strings Strings) -> &'strings str {
        use self::Type::*;
        match self {
            LeftParen => "(",
            RightParen => ")",
            LeftBrace => "{",
            RightBrace => "}",
            Comma => ",",
            Dot => ".",
            Minus => "-",
            Plus => "+",
            Semicolon => ";",
            Colon => ":",
            Slash => "/",
            Backslash => "\\",
            Star => "*",
            Underscore => "_",
            Pipe => "|",

            Bang => "!",
            BangEqual => "!=",
            Equal => "=",
            EqualEqual => "==",
            Greater => ">",
            GreaterEqual => ">=",
            Less => "<",
            LessEqual => "<=",
            Arrow => "->",

            Identifier(symbol) => strings.resolve(*symbol),
            CapitalizedIdentifier(symbol) => strings.resolve(*symbol),
            String_(symbol) => strings.resolve(*symbol),
            Float(symbol) => strings.resolve(*symbol),

            And => "and",
            Or => "or",
            Not => "not",
            If => "if",
            Then => "then",
            Else => "else",
            True => "True",
            False => "False",
            Let => "let",
            In => "in",
            As => "as",
            Exposing => "exposing",
            Module => "module",
            Import => "import",
            Type => "type",

            Comment => "[Comment]",
            Eof => "[End of file]",
        }
    }
}
