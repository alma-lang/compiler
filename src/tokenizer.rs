use std::iter::Peekable;
use std::str::CharIndices;

use crate::source::Source;
use crate::token::{self, Token};

#[derive(PartialEq, Debug)]
pub struct Error {
    pub line: u32,
    pub column: u32,
    pub message: String,
}

impl Error {
    pub fn to_string(&self, source: &Source) -> String {
        let mut msg = String::new();

        msg.push_str(&format!(
            "{}:{}:{}\n\n{}",
            source.name(),
            self.line,
            self.column,
            self.message
        ));

        msg
    }
}

enum Status {
    Reset,
    SingleToken,
    DoubleToken(token::Type),
    LineCommentToken,
    WhitespaceToken(bool),
    StringToken(bool),
    NumberToken(bool),
    IdentifierToken(bool),
}

struct State<'a> {
    status: Status,
    source: &'a Source<'a>,
    chars: Peekable<CharIndices<'a>>,
    tokens: Vec<Token>,
    errors: Vec<Error>,
    start: usize,
    start_line: u32,
    current: usize,
    current_char: char,
    line: u32,
    line_start_position: usize,
    indent: u32,
}

impl<'a> State<'a> {
    fn new(source: &'a Source) -> State<'a> {
        State {
            status: Status::SingleToken,
            tokens: vec![],
            errors: vec![],
            start: 0,
            start_line: 1,
            // current and current_char will be properly initialized on self.parse
            current: 0,
            current_char: 'a',
            line: 1,
            line_start_position: 0,
            indent: 0,
            source,
            chars: source.char_indices().peekable(),
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.chars.peek().map(|(_i, c)| *c)
    }

    fn add_token(&mut self, kind: token::Type) {
        self.push_token(Token {
            kind,
            line: self.line,
            column: self.start as u32 - self.line_start_position as u32,
            position: self.start,
            end_position: self.current + self.current_char.len_utf8(),
            indent: self.indent,
        });

        self.status = Status::Reset
    }

    fn add_error(&mut self, message: &str, use_start: bool) {
        let mut message = message.to_owned();

        let (position, line, column) = if use_start {
            (
                self.start,
                self.start_line,
                (self.start - self.line_start_position) as u32,
            )
        } else {
            (
                self.current,
                self.line,
                (self.current - self.line_start_position) as u32,
            )
        };

        message.push('\n');
        message.push('\n');
        message.push_str(
            &self
                .source
                .lines_report_at_position_with_pointer(position, None, line)
                .unwrap(),
        );

        self.errors.push(Error {
            line: self.line,
            column: column as u32,
            message,
        });

        self.status = Status::Reset;
    }

    fn push_token(&mut self, token: Token) {
        self.tokens.push(token);
    }

    fn parse_token(&mut self, ch: Option<char>) {
        use crate::token::Type::*;

        match self.status {
            Status::SingleToken => match ch {
                None => self.status = Status::Reset,
                Some(ch) => match ch {
                    '(' => self.add_token(LeftParen),
                    ')' => self.add_token(RightParen),
                    '{' => self.add_token(LeftBrace),
                    '}' => self.add_token(RightBrace),
                    ',' => self.add_token(Comma),
                    '.' => self.add_token(Dot),
                    '+' => self.add_token(Plus),
                    ';' => self.add_token(Semicolon),
                    ':' => self.add_token(Colon),
                    '*' => self.add_token(Star),
                    '\\' => self.add_token(Backslash),

                    '-' => match self.peek() {
                        Some('>') => self.status = Status::DoubleToken(Arrow),
                        _ => self.add_token(Minus),
                    },

                    '!' => match self.peek() {
                        Some('=') => self.status = Status::DoubleToken(BangEqual),
                        _ => self.add_token(Bang),
                    },

                    '=' => match self.peek() {
                        Some('=') => self.status = Status::DoubleToken(EqualEqual),
                        _ => self.add_token(Equal),
                    },

                    '<' => match self.peek() {
                        Some('=') => self.status = Status::DoubleToken(LessEqual),
                        _ => self.add_token(Less),
                    },

                    '>' => match self.peek() {
                        Some('=') => self.status = Status::DoubleToken(GreaterEqual),
                        _ => self.add_token(Greater),
                    },

                    '/' => match self.peek() {
                        Some('/') => self.status = Status::LineCommentToken,
                        _ => self.add_token(Slash),
                    },

                    '"' => self.status = Status::StringToken(false),

                    ' ' | '\n' => {
                        self.status = Status::WhitespaceToken(false);
                        self.parse_token(Some(ch));
                    }

                    '\r' | '\t' => self.status = Status::Reset,

                    ch if ch.is_digit(10) => {
                        self.status = Status::NumberToken(false);
                        self.parse_token(Some(ch));
                    }

                    ch if ch.is_alphabetic() => {
                        self.status = Status::IdentifierToken(true);
                        self.parse_token(Some(ch));
                    }

                    ch => self.add_error(&format!("Unexpected character '{}'.", ch), false),
                },
            },

            Status::DoubleToken(token_type) => self.add_token(token_type),

            Status::LineCommentToken => match ch {
                Some('\n') | None => self.add_token(Comment),
                _ => (),
            },

            Status::WhitespaceToken(line_started) => {
                match ch {
                    Some(' ') => {
                        if line_started {
                            self.indent = self.indent + 1;
                        }
                        self.status = Status::WhitespaceToken(line_started);
                    }

                    Some('\n') => {
                        self.status = Status::WhitespaceToken(true);
                        self.line = self.line + 1;
                        self.indent = 0;
                        self.line_start_position = self.current + 1;
                    }

                    _ => panic!(
                        "Got to the whitespace tokenizer without a valid space or newline char."
                    ),
                }

                match self.peek() {
                    Some('\n') | Some(' ') => (),
                    _ => self.status = Status::Reset,
                }
            }

            Status::StringToken(prev_was_backslash) => match ch {
                Some('\\') => self.status = Status::StringToken(true),
                Some('"') if !prev_was_backslash => self.add_token(String_),
                None => self.add_error("Unclosed string.", true),
                _ => {
                    if prev_was_backslash {
                        self.status = Status::StringToken(false);
                    }
                }
            },

            Status::NumberToken(seen_dot) => match ch {
                Some('.') => {
                    if seen_dot {
                        self.add_error("Multiple dots while parsing a number.", false);
                    } else {
                        self.status = Status::NumberToken(true);
                    }

                    match self.peek() {
                        Some(c) if c.is_digit(10) => (),
                        _ => self.add_error("Expected more digits after a dot in a number.", false),
                    }
                }

                Some(c) if c.is_digit(10) => match self.peek() {
                    // Continue munching until there are no numbers or dot
                    Some('.') => (),
                    Some(c) if c.is_digit(10) => (),

                    // Anything else we find ends the number token
                    _ => self.add_token(Float),
                },

                _ => panic!("Got to the number tokenizer without a valid number digit."),
            },

            Status::IdentifierToken(is_start) => match ch {
                Some(c) if (is_start && c.is_alphabetic()) || is_identifier_rest(c) => {
                    let is_start = false;
                    self.status = Status::IdentifierToken(is_start);

                    match self.peek() {
                        Some(c) if (is_start && c.is_alphabetic()) || is_identifier_rest(c) => (),

                        // Anything else we find ends the identifier token
                        _ => self.add_token(
                            match self.source.text_at(self.start..=self.current).unwrap() {
                                "and" => And,
                                "or" => Or,
                                "not" => Not,
                                "if" => If,
                                "then" => Then,
                                "else" => Else,
                                "True" => True,
                                "False" => False,
                                "let" => Let,
                                "import" => Import,
                                "as" => As,
                                "exports" => Exports,
                                "module" => Module,
                                _ => Identifier,
                            },
                        ),
                    }
                }

                _ => panic!("Got to the identifier tokenizer without a valid number digit."),
            },

            Status::Reset => {
                self.start = self.current;
                self.status = Status::SingleToken;

                self.parse_token(ch);
            }
        }
    }

    fn parse(&mut self) {
        while let Some((i, c)) = self.chars.next() {
            if let Status::Reset = &self.status {}

            self.current = i;
            self.current_char = c;

            self.parse_token(Some(c));
        }

        self.parse_token(None);

        self.tokens.push(Token {
            kind: token::Type::Eof,
            line: self.line,
            // If the last char is a \n, then line_start_position may be bigger than the last \n
            // position. Default to column 0 then.
            column: self
                .source
                .len()
                .checked_sub(self.line_start_position)
                .unwrap_or(0) as u32,
            position: self.source.len(),
            end_position: self.source.len(),
            indent: self.indent,
        });
    }
}

pub fn parse<'a>(source: &Source<'a>) -> Result<Vec<Token>, Vec<Error>> {
    let mut tokenizer = State::new(&source);

    tokenizer.parse();

    if tokenizer.errors.len() > 0 {
        Err(tokenizer.errors)
    } else {
        Ok(tokenizer.tokens)
    }
}

fn is_identifier_rest(c: char) -> bool {
    c.is_alphanumeric() || c == '_'
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::source::Source;
    use crate::token::{Token, Type::*};
    use pretty_assertions::assert_eq;

    #[test]
    fn test_scan_tokens() {
        let tests = vec![
            (
                "".to_owned(),
                Ok(vec![Token {
                    kind: Eof,
                    position: 0,
                    end_position: 0,
                    line: 1,
                    indent: 0,
                    column: 0,
                }]),
            ),
            (
                "123".to_owned(),
                Ok(vec![
                    Token {
                        kind: Float,
                        position: 0,
                        end_position: 3,
                        line: 1,
                        indent: 0,
                        column: 0,
                    },
                    Token {
                        kind: Eof,
                        position: 3,
                        end_position: 3,
                        line: 1,
                        indent: 0,
                        column: 3,
                    },
                ]),
            ),
            (
                "123.345".to_owned(),
                Ok(vec![
                    Token {
                        kind: Float,
                        position: 0,
                        end_position: 7,
                        line: 1,
                        indent: 0,
                        column: 0,
                    },
                    Token {
                        kind: Eof,
                        position: 7,
                        end_position: 7,
                        line: 1,
                        indent: 0,
                        column: 7,
                    },
                ]),
            ),
            (
                "123.sd".to_owned(),
                Err(vec![Error {
                    line: 1,
                    column: 3,
                    message: "Expected more digits after a dot in a number.

  1│  123.sd
   │     ↑"
                        .to_owned(),
                }]),
            ),
            (
                "123\n or \"abc\"".to_owned(),
                Ok(vec![
                    Token {
                        kind: Float,
                        position: 0,
                        end_position: 3,
                        line: 1,
                        indent: 0,
                        column: 0,
                    },
                    Token {
                        kind: Or,
                        position: 5,
                        end_position: 7,
                        line: 2,
                        indent: 1,
                        column: 1,
                    },
                    Token {
                        kind: String_,
                        position: 8,
                        end_position: 13,
                        line: 2,
                        indent: 1,
                        column: 4,
                    },
                    Token {
                        kind: Eof,
                        position: 13,
                        end_position: 13,
                        line: 2,
                        indent: 1,
                        column: 9,
                    },
                ]),
            ),
            (
                "123\n or &\"abc\"".to_owned(),
                Err(vec![Error {
                    line: 2,
                    column: 4,
                    message: "Unexpected character \'&\'.

  1│  123
  2│   or &\"abc\"
   │      ↑"
                        .to_owned(),
                }]),
            ),
            (
                "\"asdf\\\"asdf\"".to_owned(),
                Ok(vec![
                    Token {
                        kind: String_,
                        position: 0,
                        end_position: 12,
                        line: 1,
                        indent: 0,
                        column: 0,
                    },
                    Token {
                        kind: Eof,
                        position: 12,
                        end_position: 12,
                        line: 1,
                        indent: 0,
                        column: 12,
                    },
                ]),
            ),
        ];

        for (code, expected) in tests {
            let result = parse(&Source::new_orphan(&code));
            assert_eq!(result, expected, "\nCode:\n{:?}", &code);
        }
    }
}
