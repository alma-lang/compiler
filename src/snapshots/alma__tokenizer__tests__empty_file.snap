---
source: src/tokenizer.rs
assertion_line: 427
expression: "tokenize(\"\")"
---
Ok(
    {
        Index(
            0,
        ): Token {
            kind: Eof,
            indent: 0,
            start: 0,
            end: 0,
            line: 1,
            column: 0,
        },
    },
)
